src/compiler/crystal/semantic/bindings.cr

```diff
@@ -1,7 +1,29 @@
 module Crystal
+  class ASTNode
+    # A node whose type can be frozen.
+    # When frozen, trying to change the node's type to something
+    # that doesn't match `freeze_type` will produce a compilation error.
+    # This is used for checking a method's return type, a block's
+    # expected type, instance and class vars expected types,
+    # and local vars with a type declaration.
+    module Freezable
+      property freeze_type : Type?
```

Comment:
I'm wondering if this even warrants a dedicated module. There's little benefit from that (at least for now, maybe you're already planning something else with it?) and we could just add the property directly to each class.

Reply:
Makes sense. Especially if the definition of `freeze_type` might change for some nodes (like `Def`).